### PR TITLE
implemented DoCapture PayPal method

### PR DESCRIPTION
**Pull request by toooni**

### CODE REVIEW

File: src/Payum/Paypal/ExpressCheckout/Nvp/Tests/Action/Api/DoCaptureActionTest.php [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+use Payum\Paypal\ExpressCheckout\Nvp\Request\Api\DoCapture;
+
+class DoCaptureActionTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @test
+     */
+    public function shouldBeSubClassOfBaseApiAwareAction()
+    {
+        $rc = new \ReflectionClass('Payum\Paypal\ExpressCheckout\Nvp\Action\Api\DoCaptureAction');
```
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Tests/Action/Api/DoCaptureActionTest.php]: use ::class constant

File: src/Payum/Paypal/ExpressCheckout/Nvp/Tests/Action/Api/DoCaptureActionTest.php [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+        $action->execute($request);
+    }
+
+    /**
+     * @test
+     */
+    public function shouldCallApiDoCaptureMethodWithExpectedRequiredArguments()
+    {
+        $testCase = $this;
```
Comment (makasim) [inline src/Payum/Paypal/ExpressCheckout/Nvp/Tests/Action/Api/DoCaptureActionTest.php]: not needed, you can use $this inside closure

**Comment by makasim** [comment]
What is your use case for this doCapture?

**Comment by makasim** [comment]
I guess the authorize step is missing

**Comment by toooni** [comment]
Yeah, sorry, i forgot the Api.php




The PR should be up to date now.. 


-- 

weekend4two.com / Invit Travel GmbH

Toni Rudolf

Managing Partner




Invit Travel GmbH

Hardeggerstrasse 30

CH-3008 Bern




phone CH / +41 (0)31 528 10 11

phone AT / +43 (0)7208 84 201

mail / toni.rudolf@weekend4two.com

web / www.weekend4two.com

**Comment by makasim** [comment]
I dont get the big picture, to capture the payment you have to first authorize it by setting    
`PAYMENTREQUEST_0_PAYMENTACTION=Authorization`. Do you use it with captureaction? 

https://developer.paypal.com/docs/classic/express-checkout/ht_ec-singleAuthPayment-curl-etc/

**Comment by toooni** [comment]
I will do the changes to the tests in the afternoon. 
doCapture Captures an authorized payment. In our example: We just authorize our Payments using "PAYMENTREQUEST_0_PAYMENTACTION=Authorization", after everything attached to this payment gets confirmed our System does execute the doCapture action to finish this payment.

**Comment by toooni** [comment]
I've updated the PR

**Comment by makasim** [comment]
yeah. looks good. 

I am going to hard code  PAYMENTREQUEST_0_PAYMENTACTION=Sale in Capture action, cuz I think  allowing to authorize using capture request is odd behavior. Instead I'd like to introduce new AuthorizeAction which does exactly same thing as CaptureAction but sets  `PAYMENTREQUEST_0_PAYMENTACTION=Authorization`.

Current CaptureAction should able to differ new capture payment and previously authorized ones. and act according to it.

I'll do it asap, your PR is merged with changes I described above. Maybe you able to do them?

**Comment by toooni** [comment]
You're right. This may be a bit confusing for some people.. I'll do the implementation of an AuthorizeAction and limit the CaptureAction to the Sale Action.

**Comment by toooni** [comment]
I am not sure anymore if it's the best solution to make another action for Authorization, because there is a third PAYMENTTYPE called Order, which then also might be handled separately. Don't you think it might be better to merge this PR as it is, but make a doc entry describing the case when using Authorization?

**Pull request merged**